packages/xforms-engine/src/instance/children/normalizeChildInputs.ts

```diff
@@ -0,0 +1,355 @@
+}
+
+type EditModeMetaParent = EditModeInstanceDescendant & MetaParent;
+
+const isEditModeMetaParent = (parent: MetaParent): parent is EditModeMetaParent => {
+	return isEditModeInstance(parent.rootDocument);
+};
+
+interface LeafNodeChildInput extends InstanceNodeChildInput {
+	readonly instanceNodes: readonly [StaticLeafElement];
+}
+
+const isLeafNodeChildInput = (input: InstanceNodeChildInput): input is LeafNodeChildInput => {
+	const { instanceNodes } = input;
+	const [instanceNode, ...rest] = instanceNodes;
+
+	return instanceNode != null && rest.length === 0 && instanceNode.isLeafElement();
+};
+
+interface BaseMetaLeafChildLookupResult {
+	readonly currentIndex: number | null;
+	readonly childInput: LeafNodeChildInput | null;
+}
+
+type FoundLookupResult<T> = {
+	[K in keyof T]: NonNullable<T[K]>;
+};
+
+type NotFoundLookupResult<T> = {
+	[K in keyof T]: null;
+};
+
+type MetaLeafChildLookupResult =
+	| FoundLookupResult<BaseMetaLeafChildLookupResult>
+	| NotFoundLookupResult<BaseMetaLeafChildLookupResult>;
+
+const findMetaLeafChild = (
+	parent: MetaParent,
+	inputs: readonly InstanceNodeChildInput[],
+	localName: string
+): MetaLeafChildLookupResult => {
+	const metaName = parent.definition.qualifiedName satisfies MetaName;
+	const namespaceURI = metaName.namespaceURI.href satisfies MetaNamespaceURIValue;
+
+	const result = Array.from(inputs.entries()).find(
+		(entry): entry is [number, LeafNodeChildInput] => {
+			const [, childInput] = entry;
+
+			if (!isLeafNodeChildInput(childInput)) {
+				return false;
+			}
+
+			const [instanceNode] = childInput.instanceNodes;
+			const { qualifiedName } = instanceNode;
+
+			return (
+				qualifiedName.localName === localName && qualifiedName.namespaceURI?.href === namespaceURI
+			);
+		}
+	);
+
+	if (result == null) {
+		return {
+			currentIndex: null,
+			childInput: null,
+		};
+	}
+
+	const [currentIndex, childInput] = result;
+
+	return {
+		currentIndex,
+		childInput,
+	};
+};
+
+const getInstanceIDValue = (
+	parent: MetaParent,
+	inputs: readonly InstanceNodeChildInput[]
+): string | null => {
+	const { childInput } = findMetaLeafChild(parent, inputs, INSTANCE_ID_LOCAL_NAME);
+
+	if (childInput == null) {
+		return null;
+	}
+
+	const [instanceIDNode] = childInput.instanceNodes;
+
+	return instanceIDNode.value satisfies string;
+};
+
+type AssertStaticLeafElement = (element: StaticElement) => asserts element is StaticLeafElement;
+
+const assertStaticLeafElement: AssertStaticLeafElement = (element) => {
+	if (!element.isLeafElement()) {
+		throw new Error();
+	}
+};
+
+/**
+ * @todo this name and signature are general, for the same reasons as the TODO
+ * on {@link populateDeprecatedID}.
+ */
+const buildMetaValueElement = (
+	parent: MetaParent,
+	localName: string,
+	value: string
+): StaticLeafElement => {
+	const { qualifiedName, nodeset } = parent.definition;
+	const { namespaceURI, prefix } = qualifiedName;
+	const { root } = new StaticDocument({
```

Comment:
Note: it's a bit weird to new up a `StaticDocument` and immediately discard it in favor of its `root`. The alternatives would be:

- New up a `StaticElement` directly, referencing `parent`... as its `parent`. But this would create a hierarchical relationship which doesn't make sense! The newly created element would not be a child of its `parent`.
- Create a concept in "static-dom" similar to [`DocumentFragment`](https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment). This is compelling, but I didn't want to stray into new concepts. If we did ever do this, we'll have to make some choices about how it would interop with `xpath` (because it wouldn't be a valid context node in strict XPath semantics).